Added stop button for hello audio worklet example

duojet2ez opened the pull request:
I added a stop button to the hello audio worklet example enabling the user to stop and start the sound. 

Specifically, I used audioContext.suspend() to allow the user to pause the sound with a stop button. And I added an isModuleLoaded check to avoid redundantly calling addModule. Also added isPlaying to toggle the playing state.

duojet2ez commented:
@hoch this is the single example version of my previous PR. Just to update you on getting a test page for quick and easy inspection of the PR, I started looking into this again yesterday and it does seem possible to do with github pages + github actions. I will continue researching this over the course of the week and provide updates.

hoch commented on src/audio-worklet/basic/hello-audio-worklet/main.js:
```diff
@@ -3,24 +3,50 @@
+    oscillatorNode = new OscillatorNode(context);
+    const bypasser = new AudioWorkletNode(context, 'bypass-processor');
+    oscillatorNode.connect(bypasser).connect(context.destination);
+    oscillatorNode.start();
```
I think L13~L16 can be extracted to a different method.

`isModuleLoaded` should guard around the `addModule` method, nothing else.

hoch commented on src/audio-worklet/basic/hello-audio-worklet/main.js:
```diff
@@ -3,24 +3,50 @@
-    buttonEl.textContent = 'Playing...';
-  }, false);
+    if (!isPlaying) {
+      await startAudio(audioContext);
```
You can call `audioContext.resume()` here.

hoch commented on src/audio-worklet/basic/hello-audio-worklet/main.js:
```diff
@@ -3,24 +3,50 @@
+    oscillatorNode.connect(bypasser).connect(context.destination);
+    oscillatorNode.start();
+    isModuleLoaded = true;
+  } else context.resume();
```
To make the logic clear, please keep this call within the user interaction callback (`click`) below.

hoch commented on src/audio-worklet/basic/hello-audio-worklet/main.js:
```diff
@@ -3,24 +3,50 @@
 // found in the LICENSE file.
 
 const audioContext = new AudioContext();
+let isModuleLoaded = false;
```
You will probably need two different booleans:

`isModuleLoaded`: to guard `addModule()` call.
`isGraphReady`: to guard the node construction/connection logic.

Then the actual `click` event will only care about the context's `resume()` and `suspend()` call.

I believe this is the most efficient way.

hoch commented on src/audio-worklet/basic/hello-audio-worklet/main.js:
```diff
@@ -3,24 +3,50 @@
+    }
+  });
+
+  buttonEl.addEventListener('mouseenter', () => {
```
I believe L41~L51 is not necessary. Hovering action is not visible on the mobile platform either.

hoch reviewed (changes requested):
Thanks for your patience, and I am hopeful that we can land this pretty soon.

duojet2ez commented:
Thanks for the feedback!! 

- I extracted the following lines mentioned and put them in a loadGraph function:

  oscillatorNode = new OscillatorNode(context);
  const bypasser = new AudioWorkletNode(context, 'bypass-processor');
  oscillatorNode.connect(bypasser).connect(context.destination);
  oscillatorNode.start();

- created a bool isGraphReady to serve as a guard for graph connections in startAudio function
- removed unnecessary lines 41-51
- audioContext.resume() placed in click event 
- Click event should now only care about resume and suspend

hoch reviewed (approved):
LGTM

It looks like there's a conflict, so once you resolve it I'll merge the PR.

duojet2ez commented:
Cool I think that should resolve the conflict!
